src/diffusers/models/attention.py

```diff
@@ -339,13 +346,31 @@ def forward(self, hidden_states):
 
         scale = 1 / math.sqrt(self.channels / self.num_heads)
 
+        if encoder_hidden_states is not None:
+            encoder_hidden_states = encoder_hidden_states.transpose(1, 2)
+            context_key_proj = self.context_key(encoder_hidden_states)
+            context_value_proj = self.context_value(encoder_hidden_states)
+
         query_proj = self.reshape_heads_to_batch_dim(query_proj)
         key_proj = self.reshape_heads_to_batch_dim(key_proj)
         value_proj = self.reshape_heads_to_batch_dim(value_proj)
 
+        if encoder_hidden_states is not None:
+            context_key_proj = self.reshape_heads_to_batch_dim(context_key_proj)
+            context_value_proj = self.reshape_heads_to_batch_dim(context_value_proj)
+            key_proj = torch.concat([context_key_proj, key_proj], dim=1)
+            value_proj = torch.concat([context_value_proj, value_proj], dim=1)
+
+        if attention_mask is not None:
+            attention_mask = F.pad(attention_mask, (0, query_proj.shape[1]), value=0.0)
+            attention_mask = attention_mask.repeat_interleave(self.num_heads, dim=0)
+            attention_mask = attention_mask.unsqueeze(1)
+
         if self._use_memory_efficient_attention_xformers:
             # Memory efficient attention
-            hidden_states = xformers.ops.memory_efficient_attention(query_proj, key_proj, value_proj, attn_bias=None)
+            hidden_states = xformers.ops.memory_efficient_attention(
+                query_proj, key_proj, value_proj, attn_bias=attention_mask
+            )
```

Comment:
Note that for karlo, xformers doesn't have an operator that supports these inputs. 

The UX for when xformers can't find a supported operator isn't great and it just dumps the tensors to the console

I would assume it's due to the dimensionality of the appended key and value matrices and/or the attention mask not being one of the pre-baked classes https://facebookresearch.github.io/xformers/components/ops.html#xformers.ops.AttentionMask


We could do some more digging here and improve the UX around either error messages when there's not a supported operator or potentially falling back to non-xformers attention. However, this should still be backwards compatible with previous usage of this block and I would recommend doing that work in a follow up PR